app/api/server/v1/voip/omnichannel.ts

```diff
@@ -0,0 +1,94 @@
+import { Match, check } from 'meteor/check';
+
+import { API } from '../../api';
+import { Users } from '../../../../models/server/index';
+import { hasPermission } from '../../../../authorization/server/index';
+import { LivechatVoip } from '../../../../../server/sdk';
+
+API.v1.addRoute('omnichannel.agent.extension', { authRequired: true }, {
+	// Get the extensions associated with the agent passed as request params.
+	get() {
+		if (!hasPermission(this.userId, 'query-agent-extension-association')) {
+			return API.v1.unauthorized();
+		}
+		check(this.requestParams(), Match.ObjectIncluding({
+			username: String,
+		}));
+		const user = Users.findOneByAgentUsername(this.requestParams().username, {
+			fields: { _id: 1 },
+		});
+		if (!user) {
+			return API.v1.notFound();
+		}
+		const extension = Users.getExtension(user._id, {
+			fields: { _id: 1, username: 1, extension: 1 },
+		});
+		if (!extension) {
+			return API.v1.notFound();
+		}
+		return API.v1.success(...extension);
+	},
+
+	// Create agent-extension association.
+	post() {
+		if (!hasPermission(this.userId, 'managage-agent-extension-association')) {
+			return API.v1.unauthorized();
+		}
+		check(this.bodyParams, {
+			username: String,
+			extension: String,
+		});
+		const user = Users.findOneByAgentUsername(this.bodyParams.username, {
+			fields: { _id: 1, username: 1 },
+		});
+		if (!user) {
+			return API.v1.notFound();
+		}
+		Users.setExtension(user._id, this.bodyParams.extension);
+		return API.v1.success();
+	},
+	delete() {
```

Comment:
@renatobecker @KevLehman and @murtaza98 I have added delete API. This delete API will simply delete the extension for the user. So we do not have a trail of the past extensions atleast in the user document.

Nevertheless, we can either keep this trail in the server side or client side logging. We need to design the fields which needs to be logged for auditing purpose. 

Alternatively we can maintain the extension array and it will contain 
extension: [
    {
     _id: <Extension number>
    isActive: <boolean>
    },
...
]

We will not delete the extension. on Deletion we will just make isActive false.
While creating, we shall just push a new element in the array and set isActive = true.

Reasoning about this : On a sunny day, a delete should just delete the element but when it is with any telephony systems certain records are necessary to be maintained. e.g Call detail records **(CDR)** are necessary for GDPR. Some countries like India has strict toll compliance guidelines (And hence we do not yet have publicly sold VoIP trunks). They too need CDR. We can argue that we are not a pbx because maintaining the CDR is the responsibility of the pbx (SIP server). Nevertheless, we need some information that can trace an individual agent if something goes wrong. So if we simply delete the extension, there is no tracing of an individual thru the database. We can definitely trace back thru logs and login/logout time of an agent but that would need a multilevel redirection of data.

The long story short,  we need to decide on what we need to maintain for the audit trail and I would like to hear your thoughts.

Reply:
If such a level of traceability is something that's required, then we certainly will have to create a new collection to store all these extensions. 
Amol, I'm a little concerned about storing the trace info on the user's document because I'm worried about increasing the size of the user's document in case an agent's extension changes many times. Hence, I'm suggesting moving this trace data into another collection.

Following is my proposal for the new collection
```
{
	_id: '123', 
	extension: '123123' // note, this wouldn't be required incase we're sure that all extension will be unique. hen we can maintain this data directly within _id
	history: [
		{
			userId: 'current-active-user for this extension',
			validFrom: '21st Oct'
		},
		{
			userId: 'xyz',
			validFromTs: '17th Oct',
			validToTs: '20th Oct'
		},
		{
			userId: 'abc',
			validFromTs: '1th Sept',
			validToTs: '10th Oct'
		},
	]
}
```

Then in the user's collection, we can just store the extension ID of the corresponding extension doc

Reply:
Hello @murtaza98 while its a good design, I feel that it would be an overkill to utilise this for extension. The reason is as follows :
1. Extensions are going to remain fairly constant. The chances that the user extension may change for are
   a. PBX changes the extension partitions.
   b. Agent becomes non-voip agent.   
2. validFromTs-validToTs may not be needed as this can be referred from logs.

Furthermore the problem of traceability does not end here. What happens if the agent gets deleted? (Do we hard delete the agent today?).

More and more we discuss about this, I think we need to device an external mechanism for traceability.

But I am open to implement the design you suggested if that is the optimal way. @renatobecker what do you think about the modification / deletion frequency for the agent's extension?

Reply:
Hi @amolghode1981 

If the extension data is always short, then I agree with storing it directly within users collection. 

Regarding your second point of **validFromTs-validToTs**, here are you referring to Rocket.chat server logs or some other server logs. If it's Rocket.Chat logs, then  I wouldn't recommend relying on it for this kind of info, if it's that valuable, because logs are easily purged and so this info can easily get lost. Hence I'd recommend storing it in DB instead

Next this question
> What happens if the agent gets deleted? (Do we hard delete the agent today?).

Yes, we indeed hard-delete a user right now. So if this kind of traceability is required, then we'd indeed have to override that logic by somehow preventing a user to be deleted.

Reply:
@murtaza98 Thanks for the explanations. Few more points from my side 
1. Extension is going remain fairly simple. At the most it will be an object with 2 fields, 
{
  extension: <allocated extension>,
 server: <server-id from the voip server table>
},
2. But above said scheme too is needed if we ever support multiple VoIP servers on our platform. So theoretically it will be possible, but we may not ever take that way. 
3. I am referring rocket.chat logs here, we also should make sure that the client side logs too are received on the server and stored. I think we shall address the traceability issue at higher level because we need to address more things around it. 
@renatobecker what do you think?

Reply:
I didn't get what's exactly the meaning of traceability here: if the idea is to design a way to know who was the agent/user who was using an extension, I assume that the extension number will be stored in the room document, right? We need to store VoIP data/context inside the room, I mean, relevant information and then we'll have traceability because we'll know what was the extension number of each call and who was the agent associated. If I'm missing anything, please lemme know.
Also, I think we should have an extension history collection to store associations (extension x user) and, also, make it available on the UI, it's definitely something useful. 
BTW: We should bring this discussion to open server and discuss it internally as we're talking about business rules.

Reply:
@renatobecker @KevLehman @murtaza98 Resolving this comment as we have agreed upon creating a separate collection for maintaining Voip activity **Omnichannel call data record OCDR**. With this approach, user document will contain a simple extension string. The traceability will be possible via looking at the OCDR collection.